sabnzbd/dirscanner.py

```diff
@@ -115,99 +109,169 @@ def save(self):
+        res, _ = sabnzbd.nzbparser.add_nzbfile(path, catdir=catdir, keep=False)
+        if res < 0:
+            # Retry later, for example when we can't read the file
+            self.suspected[path] = stat_tuple
+        elif res == 0:
+            self.error_reported = False
+        else:
+            self.ignored[path] = 1
+
+    def scan(self):
+        """Schedule a scan of the watched folder"""
+        if not self.loop:
+            return
+
+        if not (dirscan_dir := self.dirscan_dir):
+            return
+
+        asyncio.run_coroutine_threadsafe(self.scan_async(dirscan_dir), self.loop)
+
+    async def scan_async(self, dirscan_dir: str):
+        """Do one scan of the watched folder"""
+        async with self.lock:
+            if sabnzbd.PAUSED_ALL:
+                return
+
+            files: Set[str] = set()
+            futures: Set[asyncio.Task] = set()
+
+            for path, catdir, stat_tuple in self.get_suspected_files(dirscan_dir):
+                files.add(path)
+
+                if path in self.ignored or path in self.suspected:
+                    continue
+
+                if stat_tuple.st_size > 0:
+                    futures.add(asyncio.create_task(self.when_stable_add_nzbfile(path, catdir, stat_tuple)))
+                    await asyncio.sleep(0)
 
             # Remove files from the bookkeeping that are no longer on the disk
-            clean_file_list(self.ignored, folder, files)
-            clean_file_list(self.suspected, folder, files)
-
-        if not self.busy:
-            self.busy = True
-            dirscan_dir = self.dirscan_dir
-            if dirscan_dir and not sabnzbd.PAUSED_ALL:
-                run_dir(dirscan_dir, None)
-
-                try:
-                    dirscan_list = os.listdir(dirscan_dir)
-                except OSError:
-                    if not self.error_reported:
-                        logging.error(T("Cannot read Watched Folder %s"), filesystem.clip_path(dirscan_dir))
-                        self.error_reported = True
-                    dirscan_list = []
-
-                cats = config.get_categories()
-                for dd in dirscan_list:
-                    dpath = os.path.join(dirscan_dir, dd)
-                    if os.path.isdir(dpath) and dd.lower() in cats:
-                        run_dir(dpath, dd.lower())
-            self.busy = False
+            # Wait for the paths found in this scan to finish
+            await asyncio.gather(clean_file_list(self.ignored, files), clean_file_list(self.suspected, files), *futures)
+
+    async def scanner(self):
+        """Periodically scan the directory and add NZB files to the queue"""
+        while True:
+            if not (dirscan_speed := self.dirscan_speed):
+                break
+
+            if not (dirscan_dir := self.dirscan_dir):
+                break
+
+            try:
+                await self.scan_async(dirscan_dir)
+            except:
+                pass
```

Comment:
I wasn't sure what to do with this, the thread never ends because of `loop.run_forever` so I need we need to be certain an unhandled exception (it catches OSError) won't stop the next scan.

There's also some code 'somewhere' that checks `is_alive` and restarts the thread (although I'm not sure if it actaully does) but again as far as I know that'll never happen.

Perhaps it should at least log it? logging.info or logging.debug?